File: contrib/sdk-bindings/kotlin/demo/app/src/test/kotlin/org/example/TestForging.kt
/* SPDX-FileCopyrightText: 2025 Functori <contact@functori.com>
 *
 * SPDX-License-Identifier: MIT
 */

package org.example.tezos

import kotlin.test.Test
import kotlin.test.assertContentEquals

class ForgingTest {

    @Test fun messageForging() {
        val msg = "message"
        val rawMsg = forgeMessage(msg)
        val expectedBytes = byteArrayOf(
            0x05, 0x01, 0x00, 0x00, 0x00, 0x07,
            'm'.code.toByte(), 'e'.code.toByte(), 's'.code.toByte(),
            's'.code.toByte(), 'a'.code.toByte(), 'g'.code.toByte(),
            'e'.code.toByte()
        )
        assertContentEquals(rawMsg, expectedBytes)
    }

    /*
    octez-codec encode "023-PtSeouLo.operation.contents" from '{
      "kind": "reveal",
      "source": "tz3NGKzVp8ezNnu5qx8mY3iioSUXKfC1d8Yc",
      "fee": "1000",
      "counter": "34",
      "gas_limit": "1",
      "storage_limit": "16",
      "public_key": "p2pk66qj2rKg5s6NCHprzPHheLK1dV3ZmGJbfBJEAizxsz2o3qms8vF"
    }'
    */
    @OptIn(kotlin.ExperimentalStdlibApi::class) // `hexToByteArray` is experimental
    @Test fun revealForging() {
        val publicKey = PublicKey.fromB58check("p2pk66qj2rKg5s6NCHprzPHheLK1dV3ZmGJbfBJEAizxsz2o3qms8vF")
        val source = PublicKeyHash.fromB58check("tz3NGKzVp8ezNnu5qx8mY3iioSUXKfC1d8Yc")
        val reveal = Operation.reveal(source = source, fee = 1000UL, counter = 34UL, gasLimit = 1UL, storageLimit = 16UL, publicKey = publicKey)
        val rawReveal = reveal.forge()
        val expectedBytes = "6b02153c42139fbbe509e9023bb85eac281709766070e80722011002032abac5ad6fc0fbe8e9a0beb3bbbc7481318f8b686b366ceb00ee0b3b51e40c4900".hexToByteArray()
        assertContentEquals(rawReveal, expectedBytes)
    }

    /*
    octez-codec encode "023-PtSeouLo.operation.contents" from '{
      "kind": "transaction",
      "source": "tz4Uzyxg26DJyM4pc1V2pUvLpdsR5jdyzYsZ",
      "fee": "30",
      "counter": "20",
      "gas_limit": "54",
      "storage_limit": "45",
      "amount": "0",
      "destination": "KT1WtTP1yGNfJLvBSubDV1H5z1zsKKXzAyGS",
      "parameters": {
        "entrypoint": "add_liquidity",
        "value": {
          "prim": "Pair",
          "args": [
            { "int": "93735" },
            { "prim": "Pair", "args": [
              { "int": "0" },
              { "string": "tz3c2XcRc6PhNuFwH1EhYcXtv7gBgdsx8k6J" }
            ] }
          ]
        }
      }
    }'
    */
    @OptIn(kotlin.ExperimentalStdlibApi::class) // `hexToByteArray` is experimental
    @Test fun transactionForging() {
        val source = PublicKeyHash.fromB58check("tz4Uzyxg26DJyM4pc1V2pUvLpdsR5jdyzYsZ")
        val destination = Contract.fromB58check("KT1WtTP1yGNfJLvBSubDV1H5z1zsKKXzAyGS")
        val entrypoint = Entrypoint("add_liquidity")
        // octez-client convert data '(Pair 93735 (Pair 0 "tz3c2XcRc6PhNuFwH1EhYcXtv7gBgdsx8k6J"))' from Michelson to binary
        val value = "070700a7b80b070700000100000024747a336332586352633650684e75467748314568596358747637674267647378386b364a".hexToByteArray()
        val transaction = Operation.transaction(source = source, fee = 30UL, counter = 20UL, gasLimit = 54UL, storageLimit = 45UL, amount = 0UL, destination = destination, entrypoint = entrypoint, value = value)
        val rawTransaction = transaction.forge()
        val expectedBytes = "6c03db557924e5a295652eff2c1f141d5a5b72b9cc911e14362d0001f4ab1300637efe3616ab27663ee0e4be07d80e4c00ffff0d6164645f6c697175696469747900000033070700a7b80b070700000100000024747a336332586352633650684e75467748314568596358747637674267647378386b364a".hexToByteArray()
        assertContentEquals(rawTransaction, expectedBytes)
    }

    /*
    octez-codec encode "023-PtSeouLo.operation.contents" from '{
      "kind": "origination",
      "source": "tz4F2fxv7sKQx9wyoRMteoJwZEZnV9WFU2wL",
      "fee": "72",
      "counter": "41",
      "gas_limit": "8017",
      "storage_limit": "77",
      "balance": "400",
      "delegate": "tz4F2fxv7sKQx9wyoRMteoJwZEZnV9WFU2wL",
      "script": {
        "code": [],
        "storage": {
          "prim": "unit"
        }
      }
    }'
    */
    @OptIn(kotlin.ExperimentalStdlibApi::class) // `hexToByteArray` is experimental
    @Test fun OriginationForging() {
        val source = PublicKeyHash.fromB58check("tz4F2fxv7sKQx9wyoRMteoJwZEZnV9WFU2wL")
        val origination = Operation.origination(source = source, fee = 72UL, counter = 41UL, gasLimit = 8017UL, storageLimit = 77UL, balance = 400UL, delegate = source)
        val rawOrigination = origination.forge()
        val expectedBytes = "6d03421585a37470f6cbba7aeb70d11bc40b2b23dd684829d13e4d9003ff03421585a37470f6cbba7aeb70d11bc40b2b23dd6800000005020000000000000002036c".hexToByteArray()
        assertContentEquals(rawOrigination, expectedBytes)
    }

    /*
    octez-codec encode "023-PtSeouLo.operation.contents" from '{
      "kind": "delegation",
      "source": "tz4Uzyxg26DJyM4pc1V2pUvLpdsR5jdyzYsZ",
      "fee": "0",
      "counter": "54",
      "gas_limit": "82",
      "storage_limit": "107006"
    }'
    */
    @OptIn(kotlin.ExperimentalStdlibApi::class) // `hexToByteArray` is experimental
    @Test fun removeDelegateForging() {
        val source = PublicKeyHash.fromB58check("tz4Uzyxg26DJyM4pc1V2pUvLpdsR5jdyzYsZ")
        val delegation = Operation.delegation(source = source, fee = 0UL, counter = 54UL, gasLimit = 82UL, storageLimit = 107006UL, delegate = null)
        val rawDelegation = delegation.forge()
        val expectedBytes = "6e03db557924e5a295652eff2c1f141d5a5b72b9cc91003652fec30600".hexToByteArray()
        assertContentEquals(rawDelegation, expectedBytes)
    }

    /*
    octez-codec encode "023-PtSeouLo.operation.unsigned" from '{
      "branch": "BMUjWA9SovWoMPzvjXyxzu1xZo8P3zNZHTKXivGRFFwaJBb856Q",
      "contents": [
        {
          "kind": "reveal",
          "source": "tz4TWB5Rmg9bqwn8wMpYRaaPax9NCtGSUN5h",
          "fee": "449",
          "counter": "444131",
          "gas_limit": "981",
          "storage_limit": "0",
          "public_key": "BLpk1wNkwwNS8QFmRfeB79DMHV9gRcg5gnGL4QqpKjqY87RLNJMwUmKhvvYSbD4Lo4omysqdnV1z",
          "proof": "BLsig9eDgfCYSwPbWQhGVDUHcnGscTqPuNTtbEwTJjjXcczUnSGNMdqdtupYtTfwzhJLpeTqnsxPzzbg7B4g3cdkky7gJZCyL53dkc4ULzFHJDi4gY9NF4EZvJ7mUZcoWpPC8sRn52ZYPs"
        },
        {
          "kind": "delegation",
          "source": "tz4TWB5Rmg9bqwn8wMpYRaaPax9NCtGSUN5h",
          "fee": "843",
          "counter": "444132",
          "gas_limit": "511",
          "storage_limit": "86",
          "delegate": "tz4TWB5Rmg9bqwn8wMpYRaaPax9NCtGSUN5h"
        },
        {
          "kind": "origination",
          "source": "tz4TWB5Rmg9bqwn8wMpYRaaPax9NCtGSUN5h",
          "fee": "686",
          "counter": "444133",
          "gas_limit": "456",
          "storage_limit": "428",
          "balance": "0",
          "delegate": "tz4TWB5Rmg9bqwn8wMpYRaaPax9NCtGSUN5h",
          "script": {
            "code": [],
            "storage": {
              "prim": "unit"
            }
          }
        },
        {
          "kind": "transaction",
          "source": "tz4TWB5Rmg9bqwn8wMpYRaaPax9NCtGSUN5h",
          "fee": "547",
          "counter": "444134",
          "gas_limit": "968",
          "storage_limit": "47",
          "amount": "0",
          "destination": "KT1NYriLXv6V5GaggQscXH5YZooeVQEmomEu",
          "parameters": {
            "entrypoint": "default",
            "value": { "int": "13" }
          }
        }
      ]
    }'
    */
    @OptIn(kotlin.ExperimentalStdlibApi::class) // `hexToByteArray` is experimental
    @Test fun operationForging() {
        val pkh = PublicKeyHash.fromB58check("tz4TWB5Rmg9bqwn8wMpYRaaPax9NCtGSUN5h")
        val pk = PublicKey.fromB58check("BLpk1wNkwwNS8QFmRfeB79DMHV9gRcg5gnGL4QqpKjqY87RLNJMwUmKhvvYSbD4Lo4omysqdnV1z")

        val reveal = Operation.reveal(
            source = pkh,
            fee = 449UL,
            counter = 444131UL,
            gasLimit = 981UL,
            storageLimit = 0UL,
            publicKey = pk,
            proof = BlsSignature.fromB58check("BLsig9eDgfCYSwPbWQhGVDUHcnGscTqPuNTtbEwTJjjXcczUnSGNMdqdtupYtTfwzhJLpeTqnsxPzzbg7B4g3cdkky7gJZCyL53dkc4ULzFHJDi4gY9NF4EZvJ7mUZcoWpPC8sRn52ZYPs")
        )
        val delegation = Operation.delegation(
            source = pkh,
            fee = 843UL,
            counter = 444132UL,
            gasLimit = 511UL,
            storageLimit = 86UL,
            delegate = pkh
        )
        val origination = Operation.origination(
            source = pkh,
            fee = 686UL,
            counter = 444133UL,
            gasLimit = 456UL,
            storageLimit = 428UL,
            balance = 0UL,
            delegate = pkh
        )
        val transaction = Operation.transaction(
            source = pkh,
            fee = 547UL,
            counter = 444134UL,
            gasLimit = 968UL,
            storageLimit = 47UL,
            amount = 0UL,
            destination = Contract.fromB58check("KT1NYriLXv6V5GaggQscXH5YZooeVQEmomEu"),
            // octez-client convert data '13' from Michelson to binary
            value = "000d".hexToByteArray()
        )

        val branch = BlockHash.fromB58check("BMUjWA9SovWoMPzvjXyxzu1xZo8P3zNZHTKXivGRFFwaJBb856Q")
        val operations = listOf(reveal, delegation, origination, transaction)

        val forgedOperation = forgeOperation(branch, operations)

        val expectedBytes = "e8a0bf2fced89c606cbd506141c5db242c10d583f3e8db55b62e4b796a930a8b6b03caea906857692736dd9d8440bab0638cc57d15a8c103e38d1bd5070003acb5946d66a3a423951f864714200ba27dd58ba59e58f5b8e5579c53c4dc498cec391cb69b257889ebbb3dc52d25d1f3ff0000006083822b70fa68eceea7f18986efe9db0be816c44d4818faf7a04a9a2dfca72ea5d1d238761c941dca0a8c84979bdbe8230cf15973b0feecefc29d02117729571edb9f2ed6b618d46e23947b124914582c637aac0e3d069bf712a07bb2e7ae53546e03caea906857692736dd9d8440bab0638cc57d15a8cb06e48d1bff0356ff03caea906857692736dd9d8440bab0638cc57d15a86d03caea906857692736dd9d8440bab0638cc57d15a8ae05e58d1bc803ac0300ff03caea906857692736dd9d8440bab0638cc57d15a800000005020000000000000002036c6c03caea906857692736dd9d8440bab0638cc57d15a8a304e68d1bc8072f0001993539e91d1794f14316375d0a1d98b014ece3f700ff0000000002000d".hexToByteArray()
        assertContentEquals(forgedOperation, expectedBytes)
    }
}

File: contrib/sdk-bindings/rust/python-tests/test_forging.py
# SPDX-FileCopyrightText: 2025 Functori <contact@functori.com>
#
# SPDX-License-Identifier: MIT

from tezos import (
    forge_message,
    forge_operation,
    Operation,
    PublicKey,
    PublicKeyHash,
    BlsSignature,
    Contract,
    BlockHash,
)

def test_message_forging():
    msg = "message"
    raw_msg = forge_message(msg)
    expected_bytes = bytes([
        0x05, 0x01, 0x00, 0x00, 0x00, 0x07,
        ord('m'), ord('e'), ord('s'), ord('s'),
        ord('a'), ord('g'), ord('e')
    ])
    assert raw_msg == expected_bytes


def test_reveal_forging():
    """
    octez-codec encode "023-PtSeouLo.operation.contents" from '{
      "kind": "reveal",
      "source": "tz4HWRiWnefpPHKUmjTfJhHG74q7fEmS5euR",
      "fee": "3",
      "counter": "718",
      "gas_limit": "3679",
      "storage_limit": "25",
      "public_key": "BLpk1nco14hJwGiTzrRp2FBCxZz6ZJ19QRAoh8hyaNMeVARFyuvTGZQgmrPTT5ZkXfS5yKuEupPN",
      "proof": "BLsigAy1Lc4y9kagjUjPFRb5avizhuXSYQyeSsGAsWbiLBwAEHZvQMxrNrH11Bm9SfHpBbCb9RjnucX2FEZbTZ6it7Z7UGFf3mo7uUdi8V2DX5SmaqqwyzyD12E3rUKxgwHkaSyEiMKcTB"
    }'
    """
    public_key = PublicKey.from_b58check("BLpk1nco14hJwGiTzrRp2FBCxZz6ZJ19QRAoh8hyaNMeVARFyuvTGZQgmrPTT5ZkXfS5yKuEupPN")
    source = PublicKeyHash.from_b58check("tz4HWRiWnefpPHKUmjTfJhHG74q7fEmS5euR")
    proof = BlsSignature.from_b58check("BLsigAy1Lc4y9kagjUjPFRb5avizhuXSYQyeSsGAsWbiLBwAEHZvQMxrNrH11Bm9SfHpBbCb9RjnucX2FEZbTZ6it7Z7UGFf3mo7uUdi8V2DX5SmaqqwyzyD12E3rUKxgwHkaSyEiMKcTB")
    reveal = Operation.reveal(
        source=source,
        fee=3,
        counter=718,
        gas_limit=3679,
        storage_limit=25,
        public_key=public_key,
        proof=proof,
    )
    raw_reveal = reveal.forge()
    expected_bytes = bytes.fromhex('6b035d4586d123b894f9d7cd1b28fff62e6b4bb8dd1103ce05df1c1903883927221f890dd4873e2296498a888d79b8020fe15b47c52547da05e73197fa21a44d87ef2ac7e69e91238031d9e531ff00000060a592dd097f55fa1df1598f577e8005413e822489bba23114cd9ed8bd2dd66891ecd52ec0ca3ff250badb9b3261f36ee9127847540ec6600e38013757c07ef249f6cf6270519a101bf143e232a1e9390952e8699a16999ef21efa267aac2e54cc')
    assert raw_reveal == expected_bytes


def test_transaction_forging():
    """
    octez-codec encode "023-PtSeouLo.operation.contents" from '{
      "kind": "transaction",
      "source": "tz3S6P2LccJNrejt27KvJRb3BcuS5vGghkP8",
      "fee": "1000000000",
      "counter": "123456",
      "gas_limit": "0",
      "storage_limit": "0",
      "amount": "1",
      "destination": "tz3hqqamVC1G22LACFoMgcJeFKZgoGMFSfSn"
    }'
    """
    source = PublicKeyHash.from_b58check("tz3S6P2LccJNrejt27KvJRb3BcuS5vGghkP8")
    destination = Contract.from_b58check("tz3hqqamVC1G22LACFoMgcJeFKZgoGMFSfSn")
    transaction = Operation.transaction(
        source=source,
        fee=1000000000,
        counter=123456,
        gas_limit=0,
        storage_limit=0,
        amount=1,
        destination=destination,
    )
    raw_transaction = transaction.forge()
    expected_bytes = bytes.fromhex('6c023f3b21e54e98db3845a2a5033d96877f7f9cea878094ebdc03c0c4070000010002ebfd1371b542831b4be730161d08885c5312e44200')
    assert raw_transaction == expected_bytes


def test_origination_forging():
    """
    octez-codec encode "023-PtSeouLo.operation.contents" from '{
      "kind": "origination",
      "source": "tz3S6P2LccJNrejt27KvJRb3BcuS5vGghkP8",
      "fee": "2",
      "counter": "603",
      "gas_limit": "59",
      "storage_limit": "109",
      "balance": "90",
      "script": {
        "code": [],
        "storage": {
          "prim": "unit"
        }
      }
    }'
    """
    source = PublicKeyHash.from_b58check("tz3S6P2LccJNrejt27KvJRb3BcuS5vGghkP8")
    origination = Operation.origination(
        source=source,
        fee=2,
        counter=603,
        gas_limit=59,
        storage_limit=109,
        balance=90,
        delegate=None
    )
    raw_origination = origination.forge()
    expected_bytes = bytes.fromhex('6d023f3b21e54e98db3845a2a5033d96877f7f9cea8702db043b6d5a0000000005020000000000000002036c')
    assert raw_origination == expected_bytes


def test_delegation_forging():
    """
    octez-codec encode "023-PtSeouLo.operation.contents" from '{
      "kind": "delegation",
      "source": "tz3hqqamVC1G22LACFoMgcJeFKZgoGMFSfSn",
      "fee": "3141",
      "counter": "24006",
      "gas_limit": "90",
      "storage_limit": "1",
      "delegate": "tz2PbzLDYrPAZS38BteBY7gqtnZfsTqHF2xu"
    }'
    """
    delegate = PublicKeyHash.from_b58check("tz2PbzLDYrPAZS38BteBY7gqtnZfsTqHF2xu")
    source = PublicKeyHash.from_b58check("tz3hqqamVC1G22LACFoMgcJeFKZgoGMFSfSn")
    delegation = Operation.delegation(
        source=source,
        fee=3141,
        counter=24006,
        gas_limit=90,
        storage_limit=1,
        delegate=delegate,
    )
    raw_delegation = delegation.forge()
    expected_bytes = bytes.fromhex('6e02ebfd1371b542831b4be730161d08885c5312e442c518c6bb015a01ff01a7b4ff0f28869d9f9c27d653c73aee41bd7fc777')
    assert raw_delegation == expected_bytes


def test_operation_forging():
    """
    jq -n '{
      "branch": "BKuGrbLSk7bBwh3x9PAXtao3bLJBtSFN5syraP4mz414BAeSxaT",
      "contents": [range(1024) | {
        "kind": "transaction",
        "source": "tz3bCBLYQUspvznkeq8cZrM1dnd1mpNqNyTF",
        "fee": "806",
        "counter": "160",
        "gas_limit": "614",
        "storage_limit": "407",
        "amount": "42",
        "destination": "tz4MFCwVMrN89ymtDsPMAXpdgug5SmGcLLJt"
      }]
    }' > /tmp/tezos-sdk-bindings-tests-many-operations.json
    octez-codec encode "023-PtSeouLo.operation.unsigned" from /tmp/tezos-sdk-bindings-tests-many-operations.json
    """
    NB_TX = 1024
    transaction = Operation.transaction(
        source=PublicKeyHash.from_b58check("tz3bCBLYQUspvznkeq8cZrM1dnd1mpNqNyTF"),
        fee=806,
        counter=160,
        gas_limit=614,
        storage_limit=407,
        amount=42,
        destination=Contract.from_b58check("tz4MFCwVMrN89ymtDsPMAXpdgug5SmGcLLJt"),
    )

    branch = BlockHash.from_b58check("BKuGrbLSk7bBwh3x9PAXtao3bLJBtSFN5syraP4mz414BAeSxaT")
    operations = [transaction] * NB_TX

    forged_operation = forge_operation(branch, operations)

    expected_bytes = bytes.fromhex(
        # octez-codec encode "operation.shell_header" from '{ "branch": "BKuGrbLSk7bBwh3x9PAXtao3bLJBtSFN5syraP4mz414BAeSxaT" }'
        "18f543ab7d89ffceddb0704543a1909fe5d29ab091379d34a04252867e40e84c" +
        # octez-codec encode "023-PtSeouLo.operation.contents" from '{
        #   "kind": "transaction",
        #   "source": "tz3bCBLYQUspvznkeq8cZrM1dnd1mpNqNyTF",
        #   "fee": "806",
        #   "counter": "160",
        #   "gas_limit": "614",
        #   "storage_limit": "407",
        #   "amount": "42",
        #   "destination": "tz4MFCwVMrN89ymtDsPMAXpdgug5SmGcLLJt"
        # }'
        "6c02a30d014fbe864d819e0efe698f1ebeb762a8dbc2a606a001e60497032a00038645200f1586a6c8d291fed4e57aa07da82ec59400" * NB_TX
    )
    assert forged_operation == expected_bytes
